packages/f/flex/xmake.lua

```diff
@@ -15,11 +15,16 @@ package("flex")
         add_urls("https://github.com/westes/flex/releases/download/v$(version)/flex-$(version).tar.gz")
     end
 
-    if is_subhost("msys") then
-        add_deps("pacman::flex")
-    end
+    on_fetch(function (package)
+        -- If pacman::flex already installed
+        if is_subhost("msys") and package:is_library() then
+            local msys_dir = os.getenv("MINGW_PREFIX")
+            local header = path.join(path.directory(msys_dir), "usr/include/FlexLexer.h")
+            os.trycp(header, package:installdir("include"))
```

Comment:
这种为啥会在 on_fetch 里面做，感觉有点 hack ，而且也没判断 opt.system ，不是应该在 on_install 里去 copy 到 installdir 么。

Reply:
1. 如果已经安装了，为啥还要去 copy 呢，直接走 内置的 fetch 到的 includedirs 不就好了。。而且走的系统库，也不会去用 package installdir
2. 如果没安装，走 on_install 里面去 copy 就好了，也不用走 on_fetch，等于重复 copy 了三次

另外，我是不建议这么搞，没法源码安装么

Reply:
> 如果已经安装了，为啥还要去 copy 呢，直接走 内置的 fetch 到的 includedirs 不就好了。。而且走的系统库，也不会去用 package installdir

`pacman::flex` 这个系统包目前就是不能开箱即用，因为需要找个位置放 `usr/include/FlexLexer.h` ，那么放在 `package:installdir("include")` 自然最好，其他包构建如果需要就可以统一用 `local includedir = flex:installdir("include")` 。

> 如果没安装，走 on_install 里面去 copy 就好了，也不用走 on_fetch，等于重复 copy 了三次

on_fetch 是在调 pacman 安装之前，哪来的多次拷贝。

> 另外，我是不建议这么搞，没法源码安装么

因为 `is_plat("windows", "mingw", "msys")` 限制了就没去走源码编译，可以改一下。

Reply:
> on_fetch 是在调 pacman 安装之前，哪来的多次拷贝。

on_fetch 里面，你没区分判断 opt.system ，也就是一次 fetch ，就会触发两次 on_fetch 调用，先找 system，再找 xmake 路径包。

- 安装之前，调用两次
- 安装之后，会重新 fetch，调用两次

总共四次

Reply:
而且我还是没明白为啥要 copy，`add_deps("pacman::flex")` 如果被安装和找到，本身就会自动添加对应的 includedirs

1. 如果 `add_deps("pacman::flex")`  返回了 `/usr/include` 路径，那么 flex/on_fetch 不管有没有 copy，用户端使用，本身就会触发 msys/gcc 的那个 issues
2. 如果 `add_deps("pacman::flex")` 没返回 `/usr/include` 路径，导致 FlexLexer.h 获取不到，那么完全可以走 private 私有包，然后 flex/on_install 里，吧系统路径下的  flex 库文件还有头文件整个 copy 进 installdir ，走完整安装不就好了么。 `add_deps("pacman::flex", {private = true})` 作为私有依赖，就完全不对用户导出了，仅仅用于内部安装。

Reply:
我一直在 xmake-repo 测试，在真正的工程 demo 才发现有很多问题。
首先包描述已经加上 `package:add("deps", "pacman::flex", {configs = {msystem = "msys"}})`
```lua
add_requires("flex", {kind = "library"})
add_packages("flex")
```
如果没装 `pacman::flex` ， 目前实现调用 pacman 安装后，会自动过滤 `/usr/include` 路径，所以肯定不会有 `includedirs` ， 但会跳过 on_load/on_install 直接执行 `on_test`。
如果已经安装了，on_load/on_install/on_test 都是直接跳过了。

除非在 `on_install` 去调用 `pacman` 安装，好像没啥办法了。

Reply:
```lua
package("flex")
    if is_subhost("msys") then
        add_deps("pacman::flex", {private = true, configs = {msystem = "msys"}})
    end

    on_install("@msys", function (package)
        print("install flex")
    end)
```

试了下，没安装的时候，on_install 不是好好的呗调用了么，没啥问题，然后安安分分在 on_install 里面去 `os.cp` copy pacman::flex 安装到系统路径的 库和头文件到 `package("flex"):installdir()` 不就好了么。。pacman::flex 也没导出 /usr/include ，就更没问题了。。 

```console
$ xmake l scripts/test.lua -vD --shallow flex
{
  "flex"
}

C:\msys64\tmp\.xmake\241125\xmake-repo\test
add local repository(local-repo): D:\projects\personal\xmake-repo ok!
local repositories:
    local-repo D:\projects\personal\xmake-repo

global repositories:
    build-artifacts https://gitee.com/xmake-mirror/build-artifacts.git main
    xmake-repo https://gitee.com/tboox/xmake-repo.git master
    builtin-repo D:\projects\personal\xmake\xmake\repository

4 repositories were found!
checking for platform ... mingw
checking for architecture ... x86_64
checking for mingw directory ... C:/msys64/mingw64
configure
{
    ndk_stdcxx = true
    network = public
    host = windows
    clean = true
    proxy_pac = pac.lua
    mingw = C:/msys64/mingw64
    arch = x86_64
    theme = default
    ccache = true
    kind = static
    mode = release
    plat = mingw
    buildir = build
}
checking for git ... ok
git rev-parse HEAD
checking for gcc ... ok
checking for gzip ... ok
checking for tar ... ok
git rev-parse HEAD
finding flex from pacman ..
checking for pacman ... ok
checking for pacman::flex ... no
installing flex from pacman ..
pacman -Sy --noconfirm --needed --disable-download-timeout flex --verbose
Root      : /
Conf File : /etc/pacman.conf
DB Path   : /var/lib/pacman/
Cache Dirs: /var/cache/pacman/pkg/
Hook Dirs : /usr/share/libalpm/hooks/  /etc/pacman.d/hooks/
Lock File : /var/lib/pacman/db.lck
Log File  : /var/log/pacman.log
GPG Dir   : /etc/pacman.d/gnupg/
Targets   : flex
:: Synchronizing package databases...
 clangarm64                                                                                426.5 KiB   180 KiB/s 00:02 [#######################################################################] 100%
 mingw32                                                                                   204.6 KiB   124 KiB/s 00:02 [#######################################################################] 100%
 mingw64                                                                                   448.2 KiB   242 KiB/s 00:02 [#######################################################################] 100%
 ucrt64                                                                                    480.3 KiB   241 KiB/s 00:02 [#######################################################################] 100%
 clang32                                                                                    36.3 KiB  20.9 KiB/s 00:02 [#######################################################################] 100%
 clang64                                                                                   466.9 KiB   255 KiB/s 00:02 [#######################################################################] 100%
 msys                                                                                      309.9 KiB   706 KiB/s 00:00 [#######################################################################] 100%
resolving dependencies...
looking for conflicting packages...

Packages (2) m4-1.4.19-2  flex-2.6.4-3

Total Installed Size:  1.92 MiB

:: Proceed with installation? [Y/n]
(2/2) checking keys in keyring                                                                                         [#######################################################################] 100%
(2/2) checking package integrity                                                                                       [#######################################################################] 100%
(2/2) loading package files                                                                                            [#######################################################################] 100%
(2/2) checking for file conflicts                                                                                      [#######################################################################] 100%
(2/2) checking available disk space                                                                                    [#######################################################################] 100%
:: Processing package changes...
(1/2) installing m4                                                                                                    [#######################################################################] 100%
(2/2) installing flex                                                                                                  [#######################################################################] 100%
:: Running post-transaction hooks...
(1/1) Updating the info directory file...
finding flex from pacman ..
checking for pacman::flex ... flex 2.6.4
{
  libfiles = {
    "/usr/lib/libfl.a"
  },
  static = true,
  version = "2.6.4",
  links = {
    "fl"
  },
  linkdirs = {
    "/usr/lib"
  }
}

  => install pacman::flex latest .. ok
install flex <--------------------------------------------------------- on_install
checking for flex ... no
checking for win_flex ... no
checking for flex ... no
{
  envs = {
    PATH = {
      "bin"
    }
  },
  license = "BSD-2-Clause",
  deps = { },
  artifacts = {
    installdir = "C:\Users\wangrunqing\AppData\Local\.xmake\packages\f\flex\2.6.4\b850a8b09fa7422f90d514607a6cc63f"
  },
  kind = "binary",
  repo = {
    url = "D:\projects\personal\xmake-repo",
    name = "local-repo",
    commit = "440e0cd27b2c5c0503f552b996aea53b7d43d094"
  },
  plat = "msys",
  pathenvs = {
    "ACLOCAL_PATH",
    "DYLD_LIBRARY_PATH",
    "LD_LIBRARY_PATH",
    "PKG_CONFIG_PATH",
    "CMAKE_PREFIX_PATH",
    "PYTHONPATH",
    "PATH"
  },
  name = "flex",
  version = "2.6.4",
  configs = {
    shared = false,
    pic = true,
    debug = false
  },
  librarydeps = { },
  arch = "x86_64",
  mode = "release"
}

flex -h
Usage: flex [OPTIONS] [FILE]...
Generates programs that perform pattern-matching on text.

Table Compression:
  -Ca, --align      trade off larger tables for better memory alignment
  -Ce, --ecs        construct equivalence classes
  -Cf               do not compress tables; use -f representation
  -CF               do not compress tables; use -F representation
  -Cm, --meta-ecs   construct meta-equivalence classes
  -Cr, --read       use read() instead of stdio for scanner input
  -f, --full        generate fast, large scanner. Same as -Cfr
  -F, --fast        use alternate table representation. Same as -CFr
  -Cem              default compression (same as --ecs --meta-ecs)

Debugging:
  -d, --debug             enable debug mode in scanner
  -b, --backup            write backing-up information to lex.backup
  -p, --perf-report       write performance report to stderr
  -s, --nodefault         suppress default rule to ECHO unmatched text
  -T, --trace             flex should run in trace mode
  -w, --nowarn            do not generate warnings
  -v, --verbose           write summary of scanner statistics to stdout
      --hex               use hexadecimal numbers instead of octal in debug outputs

Files:
  -o, --outfile=FILE      specify output filename
  -S, --skel=FILE         specify skeleton file
  -t, --stdout            write scanner on stdout instead of lex.yy.c
      --yyclass=NAME      name of C++ class
      --header-file=FILE   create a C header file in addition to the scanner
      --tables-file[=FILE] write tables to FILE

Scanner behavior:
  -7, --7bit              generate 7-bit scanner
  -8, --8bit              generate 8-bit scanner
  -B, --batch             generate batch scanner (opposite of -I)
  -i, --case-insensitive  ignore case in patterns
  -l, --lex-compat        maximal compatibility with original lex
  -X, --posix-compat      maximal compatibility with POSIX lex
  -I, --interactive       generate interactive scanner (opposite of -B)
      --yylineno          track line count in yylineno

Generated code:
  -+,  --c++               generate C++ scanner class
  -Dmacro[=defn]           #define macro defn  (default defn is '1')
  -L,  --noline            suppress #line directives in scanner
  -P,  --prefix=STRING     use STRING as prefix instead of "yy"
  -R,  --reentrant         generate a reentrant C scanner
       --bison-bridge      scanner for bison pure parser.
       --bison-locations   include yylloc support.
       --stdinit           initialize yyin/yyout to stdin/stdout
       --nounistd          do not include <unistd.h>
       --noFUNCTION        do not generate a particular FUNCTION

Miscellaneous:
  -c                      do-nothing POSIX option
  -n                      do-nothing POSIX option
  -?
  -h, --help              produce this help message
  -V, --version           report flex version
  => install flex 2.6.4 .. ok

```

Reply:
好像终于找到问题所在。
如果用户没装 `pacman::flex` ，没问题。
如果已经装了，目前 `xmake::flex` 会自己 fetch 到 `pacman::flex` 即使加了 deps ，所以没执行 `on_install` 去 copy ，除非用 `add_requires("flex", {kind = "library", system = false})`。
```console
$ xmake f -c -v -y
checking for platform ... mingw
checking for architecture ... x86_64
checking for mingw directory ... C:/msys64/mingw64
checking for git ... ok
checking for gzip ... ok
checking for tar ... ok
git rev-parse HEAD
checking for gcc ... ok
checking for pacman ... ok
checking for pacman::flex ... flex 2.6.4
checking for xmake::flex ... no
checking for emerge ... no
checking for pkg-config ... ok
checking for flex ... pacman::flex 2.6.4
$ xmake
[ 50%]: cache compiling.release src\main.cpp
error: src\main.cpp:2:10: fatal error: FlexLexer.h: No such file or directory
    2 | #include <FlexLexer.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
  > in src\main.cpp
```

Reply:
那是因为你设置成了 binary，安装好 pacman::flex 后，它已经在 usr/bin 下了，安装 flex 时候，尝试 fetch 探测 Flex 命令存在，就默认用系统的了。。这也没啥问题哈，因为你是作为 binary 装的 flex 不是库，能有直接用的 binary 当然直接运行了。

关键是为啥会默认作为 binary kind 呢

就算里面有 tools，按理也是默认 library， 然后加 bin path .

Reply:
没考虑周到吧，只能根据需求迭代。